src/transformers/models/whisper/generation_whisper.py

```diff
@@ -0,0 +1,1436 @@
+
+                if needs_fallback[i]:
+                    new_fallback_index_map.append(fallback_index_map[i])
+                    new_segment_input.append(segment_input[i])
+                    new_decoder_input_ids.append(decoder_input_ids[i])
+                    if "decoder_attention_mask" in kwargs:
+                        new_decoder_attention_mask.append(kwargs["decoder_attention_mask"][i])
+
+            fallback_index_map = new_fallback_index_map
+
+            # if no sequence needs to be run with temperature fallback, we're finished
+            if len(fallback_index_map) == 0 or fallback_idx == len(temperatures) - 1:
+                seek_sequences = seek_sequence_list
+                seek_outputs = seek_outputs_list
+                break
+
+            # if we're still in the loop, make sure that decoder_input_ids and segment inputs are tensors
+            decoder_input_ids = torch.stack(new_decoder_input_ids)
+            segment_input = torch.stack(new_segment_input)
+            if "decoder_attention_mask" in kwargs:
+                kwargs["decoder_attention_mask"] = torch.stack(new_decoder_attention_mask)
+
+        return seek_sequences, seek_outputs, should_skip, do_condition_on_prev_tokens
+
+    def _postprocess_outputs(self, seek_outputs, return_token_timestamps, generation_config):
+        if return_token_timestamps and hasattr(generation_config, "alignment_heads"):
+            num_frames = getattr(generation_config, "num_frames", None)
+            seek_outputs["token_timestamps"] = self._extract_token_timestamps(
+                seek_outputs, generation_config.alignment_heads, num_frames=num_frames
+            )
+
+        if generation_config.return_dict_in_generate:
+
+            def split_by_batch_index(values, key, batch_idx):
+                if key == "scores":
+                    return [v[batch_idx].cpu() for v in values]
+                if key == "past_key_values":
+                    # we don't save `past_key_values` as this is too costly
+                    return None
+                return values[batch_idx].cpu()
+
+            sequence_tokens = seek_outputs["sequences"]
+            seek_outputs = [
+                {k: split_by_batch_index(v, k, i) for k, v in seek_outputs.items()}
+                for i in range(sequence_tokens.shape[0])
+            ]
+        else:
+            sequence_tokens = seek_outputs
+
+        return sequence_tokens, seek_outputs
+
+    @staticmethod
+    def _need_fallback(
+        seek_sequence,
+        seek_outputs,
+        index,
+        logits_processor,
+        compression_ratio_threshold,
+        logprob_threshold,
+        no_speech_threshold,
+        vocab_size,
+        eos_token_id,
+        temperature,
+    ):
+        needs_fallback = False
+        should_skip = False
+        if compression_ratio_threshold is not None:
+            compression_ratio = WhisperGenerationMixin._retrieve_compression_ratio(seek_sequence, vocab_size)
```

Comment:
Should `_retrieve_compression_ratio` be outside the mixin OR `_need_fallback` not be a `@staticmethod`? It's awkward to call a class method from a static method, especially when they could belong to the same instance 🤔 

(The same applies to `_retrieve_avg_logprobs`)

Reply:
Hmmm I think it's fine to call a static method via the "class object"

Reply:
Ok changing to `self`